### PR TITLE
Support live config reload for ProxySQL.

### PR DESCRIPTION
**Pull request by rtnpro**
Requires:
- Secret volume with secrets.env file to be mounted at /tmp/proxysql-secrets
- Config volume with proxysql.cnf.tpl to be mounted at /tmp/proxysql-conf

If 'PROXYSQL_CONF_LIVE_RELOAD' env var is set to 'true' (default), the entrypoint.sh script checks for config changes every 'PROXYSQL_CONF_CHECK_INTERVAL' and loads config to ProxySQL runtime on config change.

The live config reloader is able to tolerate syntax errors in proxysql config, and wait for file change to fix it. On encountering config load error from config file, it stops loading config to runtime and ProxySQL keeps running with last known working config.

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -3,15 +3,74 @@ set -e
...
-	CMDARG="$@"
-fi
+function renderProxySQLCnf {
+    cat $PROXYSQL_CONF_TPL_FILE | envsubst > $PROXYSQL_CONF_FILE && \
+    echo "Rendered $PROXYSQL_CONF_TPL_FILE"
+}
 
-exec proxysql -f $CMDARG
+function loadProxySQLCnf {
+    MYSQL_CMD='mysql -h 127.0.0.1 -P 6032 -u ${proxysql_admin_username} -p${proxysql_admin_password} -D main -e "LOAD MYSQL VARIABLES FROM CONFIG; LOAD MYSQL SERVERS FROM CONFIG; LOAD MYSQL USERS FROM CONFIG; LOAD MYSQL QUERY RULES FROM CONFIG; LOAD MYSQL VARIABLES TO RUNTIME; LOAD MYSQL SERVERS TO RUNTIME; LOAD MYSQL USERS TO RUNTIME; LOAD MYSQL QUERY RULES TO RUNTIME; SAVE MYSQL VARIABLES TO DISK; SAVE MYSQL SERVERS TO DISK; SAVE MYSQL USERS TO DISK; SAVE MYSQL QUERY RULES TO DISK;"'
```
Comment (sl1pm4t) [inline entrypoint.sh:26]: Do the `proxysql_admin_username` & `proxysql_admin_password` values come from the same mounted secret volume as the rest of the config? If so, what would happen if we tried to update these values?

File: entrypoint.sh [hunk 2]
```diff
@@ -3,15 +3,74 @@ set -e
...
+    if [ "$PROXYSQL_CONF_LIVE_RELOAD" != true ]; then
+        echo "ProxySQL live config reload is disabled."
+        return
+    fi
+    echo "ProxySQL live config reload is enabled. Checking config updates every ${PROXYSQL_CONF_CHECK_INTERVAL}s..."
+
+    touch $PROXYSQL_CONF_RELOADER_LOCK
+    echo "Starting ProxySQL config reloader..."
+    m1=$(md5sum $PROXYSQL_CONF_TPL_FILE)
+    m2=$(md5sum $PROXYSQL_SECRETS_FILE)
```
Comment (kishorenc) [inline entrypoint.sh:45]: We can call these as `tpl_hash_prev` and `secrets_hash_prev`.

File: entrypoint.sh [hunk 3]
```diff
@@ -3,15 +3,74 @@ set -e
...
+
+    while true; do
+        sleep $PROXYSQL_CONF_CHECK_INTERVAL
+        echo "Checking for ProxySQL config changes..."
+        m1_=$(md5sum $PROXYSQL_CONF_TPL_FILE)
+        m2_=$(md5sum $PROXYSQL_SECRETS_FILE)
```
Comment (kishorenc) [inline entrypoint.sh:51]: We can call these as `tpl_hash` and `secrets_hash`.

File: entrypoint.sh (hunk 1 above)
Comment (rtnpro) [reply entrypoint.sh:26]: Nice point @sl1pm4t . AFAIK, updating admin credentials will require the pods to be rotated.

**Comment by rtnpro** [comment]
> Made a couple of minor comments, but I am thinking whether the entrypoint script should be calling a Python script that handles most of the logic. Bash scripts with critical logic are not easy to maintain in the long run.

@kishorenc I had also thought of running a Python script in parallel or in a sidecar. But then, the logic did not turn up to be big enough for writing a separate Python script. Also, I have seen longer and more complicated entrypoint files for other projects. So, I think that we are good here.

**Pull request merged**